docs/050-breaking-changes.rst

```diff
@@ -0,0 +1,206 @@
+  function that does not specify its visibility already.
+
+* Explicit data location for all variables of struct, array or mapping types is
+  now mandatory. This is also applied to function parameters and return
+  variables.  For example, change ``uint[] x = m_x`` to ``uint[] storage x =
+  m_x``, and ``function f(uint[][] x)`` to ``function f(uint[][] memory x)``
+  where ``memory`` is the data location and might be replaced by ``storage`` or
+  ``calldata`` accordingly.  Note that ``external`` functions require
+  parameters with a data location of ``calldata``.
+
+* Variables of contract type do not contain ``address`` members anymore in
+  order to separate the namespaces.  Therefore, it is now necessary to
+  explicitly convert values of contract type to addresses before using an
+  ``address`` member.  Example: if ``c`` is a contract, change
+  ``c.transfer(...)`` to ``address(c).transfer(...)``.
+
+* The ``address`` type  was split into ``address`` and ``address payable``,
+  where only ``address payable`` provides the ``transfer`` function.  An
+  ``address payable`` can be converted to an ``address``, but the other way
+  around is not allowed.
+
+* Conversions between ``bytesX`` and ``uintY`` of different size are now
+  disallowed due to ``bytesX`` padding on the right and ``uintY`` padding on
+  the left which may cause unexpected conversion results.  The size must now be
+  adjusted within the type before the conversion.  For example, you can convert
+  a ``bytes4`` (4 bytes) to a ``uint64`` (8 bytes) by first converting the
+  ``bytes4`` variable to ``bytes8`` and then to ``uint64``.
+
+* Using ``msg.value`` in ``non-payable`` functions (or introducing it via a
+  modifier) is disallowed as a security feature. Turn the function into
+  ``payable`` or create a new internal function for the program logic that
+  uses ``msg.value``.
+
+Example
+=======
+
+The following example shows a contract and its updated version for Solidity
+v0.5.0.
+Old version:
+
+::
+
+   pragma solidity ^0.4.24;
+
+   contract OtherContract {
+      uint x;
+      function f(uint y) external {
+         x = y;
+      }
+      function() payable external {}
+   }
+
+   contract Old {
+      OtherContract other;
+      uint myNumber;
+
+      // Function visibility not provided, not an error
+      // Function mutability not provided, not an error
+      function f(uint x) {
+         y = -3 >> 1; // y = -1 (wrong, should be -2)
+
+         do {
+            x += 1;
+            if (x > 10) continue;
+         } while (x < 11); // causes infinite loop
+
+         bool success = other.call("f"); // Call returns only a Bool
+         if (!success)
+            revert();
+         else {
+            int y; // Local variables could be declared after their use
+         }
+      }
+
+      function g(uint[] arr, bytes8 x, OtherContract otherContract) public { // No need to explicit parameter location data
```

Comment:
``No need for an explicit data location for "arr"``?

Reply:
This is the "how it used to be" contract